src/backend/storage/smgr/md.c

```diff
@@ -516,7 +516,18 @@ mdunlinkfork(RelFileNodeBackend rnode, ForkNumber forkNum, bool isRedo, char rel
 	 */
 	if (isRedo || forkNum != MAIN_FORKNUM || RelFileNodeBackendIsTemp(rnode))
 	{
+/*
+ * GPDB: Temp tables use shared buffers in Greenplum. As a result, simple unlink
+ * is not enough to return disk space to the OS immediately, because the files
+ * of a temp relation can still be opened by the bg writer process. To fix this
+ * problem, in GPDB we skip the check if the relation is temp or not, so files
+ * of the temp relation are also truncated.
+ */
+#if 0
 		if (!RelFileNodeBackendIsTemp(rnode))
+#else
+		if (true)
```

Comment:
Why in the previous solution we called do_truncate unconditionally and now again conditionally?

Reply:
Check the entire if block:

```
	/*
	 * Delete or truncate the first segment.
	 */
	if (isRedo || forkNum != MAIN_FORKNUM || RelFileNodeBackendIsTemp(rnode))
	{
/*
 * GPDB: Temp tables use shared buffers in Greenplum. As a result, simple unlink
 * is not enough to return disk space to the OS immediately, because the files
 * of a temp relation can still be opened by the bg writer process. To fix this
 * problem, in GPDB we skip the check if the relation is temp or not, so files
 * of the temp relation are also truncated.
 */
#if 0
		if (!RelFileNodeBackendIsTemp(rnode))
#else
		if (true)
#endif
		{
			/* Prevent other backends' fds from holding on to the disk space */
			ret = do_truncate(path);
		}
		else
			ret = 0;

		/* Next unlink the file, unless it was already found to be missing */
		if (ret == 0 || errno != ENOENT)
		{
			ret = unlink(path);
			if (ret < 0 && errno != ENOENT)
				ereport(WARNING,
						(errcode_for_file_access(),
						 errmsg("could not remove file \"%s\": %m", path)));
		}
	}
	else
	{
		/* Prevent other backends' fds from holding on to the disk space */
		ret = do_truncate(path);

		/* Register request to unlink first segment later */
		register_unlink(rnode);
	}
```
It now does ``` do_truncate ``` for any case. In previous solution we pulled the ``` do_truncate ``` before the if-else. But as discussed, it introduced larger delta, that we tried to avoid.